### DISCUSSION TITLE
case() to select specific table as subquery

### DISCUSSION THREAD

**Discussion by Dreamsorcerer** [sqlalchemy/sqlalchemy#11719] [Usage Questions] [answered]
I'm really struggling to figure out the code for this, and I'm hoping someone might be able to help me out.

I have 2 tables to represent UK and US terms. If the term is identical for both, then it only exists in the UK table, and I want to use it as a fallback when searching for US terms. The columns are identical and so the tables are defined as (where `_Rhyme` is a mixin defining all the columns):
```
class Rhyme(Base, _Rhyme):
    __tablename__ = "rhymes"
    id: Mapped[int] = mapped_column(primary_key=True)
    rhyme_us: Mapped["RhymeUS"] = relationship("RhymeUS", cascade="save-update, merge, delete, delete-orphan")

class RhymeUS(Base, _Rhyme):
    __tablename__ = "rhymes_us"
    id: Mapped[int] = mapped_column(sa.ForeignKey(Rhyme.id, ondelete="CASCADE"), primary_key=True)
```

i.e. If I have tables that look like:
```
Rhyme                                       RhymeUS
id         term        ph1                  id          term       ph1
----------------------------                --------------------------
1          mum         2                    1           mom        5
2          cat         2
3          lego        7                    3           legos      2
```
Then I want to query and return the results for:
(1, mom 5)
(2, cat, 2)
(3, legos, 2)
So if I filter on ph1==2, then I should get: [(2, cat, 2), (3, legos, 2)]

So, I figured I need a subquery which returns a Rhyme or RhymeUS depending on whether a RhymeUS exists for each ID, and then do the querying against that result.

This is where I've got to so far, but this still doesn't work:
```
        cols = (sa.case((RhymeUS.id != None, getattr(RhymeUS, c.name)), else_=getattr(Rhyme, c.name)).label(c.name) for c in Rhyme.__table__.columns)
        sub_query = sa.select(*cols).outerjoin(RhymeUS).subquery()
        #alias = with_polymorphic(_Rhyme, "*", aliased=True, selectable=sub_query)  # No idea how to use this, or if it will help...
        alias = aliased(Rhyme, sub_query)

        where = sa.and_(alias.ph1 == ph1, alias.syl == 1, alias.aux_syl == 0)  # Example search query
        stmt = sa.select(*(getattr(alias, c.name) for c in Rhyme.__table__.columns)).where(where)
        results = await sess.execute(stmt)
        rhymes = [Rhyme(**dict(zip((c.name for c in Rhyme.__table__.columns), r))) for r in primary.all()]
```

This seems pretty awkward, and still seems to just return the original Rhyme results, never the US versions. Or if I change the alias to RhymeUS, then it only returns the US ones and never the fallback ones.

**Comment by CaselIT** [comment]
Hi,

Do you have an sql query that returns the data that you want? If so we can help translate it.

Note that if you want to use orm instances you probably need to setup inheritance (your example seems to use a mixin), otherwise sqlalchemy has no way of returning one class of the other

**Reply by Dreamsorcerer** [reply]
I'm no SQL expert, but I think this is working correctly:
```
stmt = sa.text("""
    SELECT anon.id, anon.term, anon.syl, anon.aux_syl, anon.ph1, anon.ph1_type, anon.ph2, anon.ph2_type, anon.ph3, anon.ph3_type, anon.shrt_bgn, anon.shrt_end
    FROM (SELECT
            (CASE WHEN rhymes_us.id IS NOT NULL THEN rhymes_us.id ELSE rhymes.id END) as id,
            (CASE WHEN rhymes_us.term IS NOT NULL THEN rhymes_us.term ELSE rhymes.term END) as term,
            (CASE WHEN rhymes_us.syl IS NOT NULL THEN rhymes_us.syl ELSE rhymes.syl END) as syl,
            (CASE WHEN rhymes_us.aux_syl IS NOT NULL THEN rhymes_us.aux_syl ELSE rhymes.aux_syl END) as aux_syl,
            (CASE WHEN rhymes_us.ph1 IS NOT NULL THEN rhymes_us.ph1 ELSE rhymes.ph1 END) as ph1,
            (CASE WHEN rhymes_us.ph1_type IS NOT NULL THEN rhymes_us.ph1_type ELSE rhymes.ph1_type END) as ph1_type,
            (CASE WHEN rhymes_us.ph2 IS NOT NULL THEN rhymes_us.ph2 ELSE rhymes.ph2 END) as ph2,
            (CASE WHEN rhymes_us.ph2_type IS NOT NULL THEN rhymes_us.ph2_type ELSE rhymes.ph2_type END) as ph2_type,
            (CASE WHEN rhymes_us.ph3 IS NOT NULL THEN rhymes_us.ph3 ELSE rhymes.ph3 END) as ph3,
            (CASE WHEN rhymes_us.ph3_type IS NOT NULL THEN rhymes_us.ph3_type ELSE rhymes.ph3_type END) as ph3_type,
            (CASE WHEN rhymes_us.shrt_bgn IS NOT NULL THEN rhymes_us.shrt_bgn ELSE rhymes.shrt_bgn END) as shrt_bgn,
            (CASE WHEN rhymes_us.shrt_end IS NOT NULL THEN rhymes_us.shrt_end ELSE rhymes.shrt_end END) as shrt_end
        FROM rhymes LEFT JOIN rhymes_us ON rhymes.id = rhymes_us.id) as anon
    WHERE anon.ph1 = {} AND anon.syl = 1 AND anon.aux_syl = 0;
""".format(ph1))
results = await sess.execute(stmt)
rhymes = [Rhyme(**dict(zip((c.name for c in Rhyme.__table__.columns), r))) for r in primary.all()]
```

**Reply by CaselIT** [reply]
That likely works only if every column is not null. Assuming id is not null, then you could use it in every case condition.

As mentioned such a query will return only one of the two classes when using the orm. If that's ok then it shouldn't be too hard to convert in sqlalchemy code. 
Also maybe using union it's easier?

**Reply by gordthompson** [reply]
IMHO, one look at that select statement would convince me to try and find a solution that does not use two separate tables. Not to mention that trying to add a third option (say, [en-]IN, CA, or AU) would be … awkward.

**Reply by Dreamsorcerer** [reply]
> Assuming id is not null, then you could use it in every case condition.

Yes, sorry, I got that wrong, every case should be `rhyme_us.id IS NOT NULL`, like the original sqlalchemy attempt.

> As mentioned such a query will return only one of the two classes when using the orm. If that's ok then it shouldn't be too hard to convert in sqlalchemy code.

As long as the data matches, that's fine. They have the same columns, so the code using it will work with either class.
But, as I described, I've failed to get anything working. Printing out the statement that sqlalchemy has produced, I've realised that it's just created a simple select without the case or join statements at all. So, I've clearly done something badly wrong there.

> Also maybe using union it's easier?

I'm not clear how I'd do this. If I understand correctly, I'd just end up with a row for UK and another for US, then I'm not clear how I'd remove the duplicate UK ones.

> one look at that select statement would convince me to try and find a solution that does not use two separate tables.

Such as?

**Reply by CaselIT** [reply]
You are using polymorphic, but in this case it's not appropriate, since there is no inhericance defined in your models.

Is seems that you could simply use the snippet you provided 
```py
cols = (sa.case((RhymeUS.id != None, getattr(RhymeUS, c.name)), else_=getattr(Rhyme, c.name)).label(c.name) for c in Rhyme.__table__.columns)
query = sa.select(*cols).outerjoin(RhymeUS)
```

Then add the where to that query directly, there's no need to use a subquery.

Then you could do
```py
final = select(Rhyme).from_statement(previous_query)
```

Haven't tried it, but something like this should do the trick 

> I'm not clear how I'd do this. If I understand correctly, I'd just end up with a row for UK and another for US, then I'm not clear how I'd remove the duplicate UK ones.

Yes, sorry. That would work only if the rows were the same

**Reply by Dreamsorcerer** [reply]
That's not quite working. One thing I'm not understanding is how to specify the where query in sqlalchemy when the comparisons are on the select results.
```
        cols = [sa.case((RhymeUS.id != None, getattr(RhymeUS, c.name)), else_=getattr(Rhyme, c.name)).label(c.name) for c in Rhyme.__table__.columns]
        sub_query = sa.select(*cols).outerjoin(RhymeUS).select_from(Rhyme)
        where = sa.and_(Rhyme.ph1 == ph1, Rhyme.syl == 1, Rhyme.aux_syl == 0)
        stmt = sa.select(Rhyme).from_statement(sub_query.where(where))
```
This produces:
```
SELECT CASE WHEN (rhymes_us.id IS NOT NULL) THEN rhymes_us.id ELSE rhymes.id END AS id, CASE WHEN (rhymes_us.id IS NOT NULL) THEN rhymes_us.term ELSE rhymes.term END AS term, CASE WHEN (rhymes_us.id IS NOT NULL) THEN rhymes_us.syl ELSE rhymes.syl END AS syl, CASE WHEN (rhymes_us.id IS NOT NULL) THEN rhymes_us.aux_syl ELSE rhymes.aux_syl END AS aux_syl, CASE WHEN (rhymes_us.id IS NOT NULL) THEN rhymes_us.ph1 ELSE rhymes.ph1 END AS ph1, CASE WHEN (rhymes_us.id IS NOT NULL) THEN rhymes_us.ph1_type ELSE rhymes.ph1_type END AS ph1_type, CASE WHEN (rhymes_us.id IS NOT NULL) THEN rhymes_us.ph2 ELSE rhymes.ph2 END AS ph2, CASE WHEN (rhymes_us.id IS NOT NULL) THEN rhymes_us.ph2_type ELSE rhymes.ph2_type END AS ph2_type, CASE WHEN (rhymes_us.id IS NOT NULL) THEN rhymes_us.ph3 ELSE rhymes.ph3 END AS ph3, CASE WHEN (rhymes_us.id IS NOT NULL) THEN rhymes_us.ph3_type ELSE rhymes.ph3_type END AS ph3_type, CASE WHEN (rhymes_us.id IS NOT NULL) THEN rhymes_us.shrt_bgn ELSE rhymes.shrt_bgn END AS shrt_bgn, CASE WHEN (rhymes_us.id IS NOT NULL) THEN rhymes_us.shrt_end ELSE rhymes.shrt_end END AS shrt_end 
FROM rhymes LEFT OUTER JOIN rhymes_us ON rhymes.id = rhymes_us.id 
WHERE rhymes.ph1 = :ph1_1 AND rhymes.syl = :syl_1 AND rhymes.aux_syl = :aux_syl_1
```
If I change the where to use `cols[4] == ph1`, then it render the case statement again:
`WHERE CASE WHEN (rhymes_us.id IS NOT NULL) THEN rhymes_us.ph1 ELSE rhymes.ph1 END = :param_1 AND CASE WHEN (rhymes_us.id IS NOT NULL) THEN rhymes_us.syl ELSE rhymes.syl END = :param_2 AND CASE WHEN (rhymes_us.id IS NOT NULL) THEN rhymes_us.aux_syl ELSE rhymes.aux_syl END = :param_3`

~~And in both cases, I just get loads of Rhyme objects that are full of Nones (including the primary keys, which can't be None).~~
Edit, Nevermind, seems to be working. Just need to remove the duplicate case statements now.

**Reply by CaselIT** [reply, marked as answer]
right, maybe subquery it's indeed better here, at least the sql is more comprehensible. that should be

```py
cols = [sa.case((RhymeUS.id != None, getattr(RhymeUS, c.name)), else_=getattr(Rhyme, c.name)).label(c.name) for c in Rhyme.__table__.columns]
sub_query = sa.select(*cols).outerjoin(RhymeUS).select_from(Rhyme).subqury()
where = sa.and_(sub_query.c.ph1 == ph1, sub_query.c.syl == 1, sub_query.c.aux_syl == 0)
stmt = sa.select(Rhyme).from_statement(select(sub_query).where(where))
```

**Reply by gordthompson** [reply]
> > one look at that select statement would convince me to try and find a solution that does not use two separate tables.
> 
> Such as?

Say we have a "rhymes" table

```
id|locale|term |ph1|
--+------+-----+---+
 1|UK    |mum  |  2|
 1|US    |mom  |  5|
 2|UK    |cat  |  7|
 3|UK    |lego |  7|
 3|US    |legos|  2|
```

Using Microsoft SQL Server (which supports CTEs) we can retrieve the US terms — including the UK fallback values — with

```sql
with scored (id, locale, score, term, ph1) as (
    select 
        id, 
        locale, 
        (case when locale = 'US' then 1 when locale = 'UK' then 0 else -1 end) as score, 
        term, 
        ph1
    from rhymes
),
top_score (id, top_score) as (
    select id, max(score) as top_score
    from scored group by id
)
select s.id, s.term, s.ph1
from scored s inner join top_score t on s.id=t.id and s.score = t.top_score;
```

returning

```
id|term |ph1|
--+-----+---+
 1|mom  |  5|
 2|cat  |  7|
 3|legos|  2|
```

**Reply by Dreamsorcerer** [reply]
> right, maybe subquery it's indeed better here, at least the sql is more comprehensible. that should be

That seems to produce the same query as I did in text, which is good.
However, the `sa.select(Rhyme)` seems to manage to always select the result from the rhymes table, regardless of what the SQL says.
I can still get it work with one of my previous hacks though:
```
stmt = sa.select(sub_query).where(where)
results = await sess.execute(stmt)
rhymes = [Rhyme(**dict(zip((c.name for c in Rhyme.__table__.columns), r))) for r in results.all()]
```

**Reply by Dreamsorcerer** [reply]
> Using Microsoft SQL Server (which supports CTEs) we can retrieve the US terms — including the UK fallback values — with

Thanks, that could work, but would probably take me many more hours to get right. As I've got this working now, I'll probably leave it. It's a small project, so performance isn't likely to be a concern (and I'll be using page caching, so the most common results won't need refetching).

If there's an easy way to clean up that last hack, I'll take it. Otherwise, I'll leave it like this.

**Reply by CaselIT** [reply]
> However, the sa.select(Rhyme) seems to manage to always select the result from the rhymes table, regardless of what the SQL says.

Strange. @zzzeek is there a way to convince the orm to map the columns on names and not on lineage? the query is this one https://github.com/sqlalchemy/sqlalchemy/discussions/11719#discussioncomment-10307758

**Reply by zzzeek** [reply]
without reading any of this thread as I'm old and tired, you can take any query and map it on a class matching by name using [aliased](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#sqlalchemy.orm.aliased) with [adapt_on_names](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#sqlalchemy.orm.aliased.params.adapt_on_names).  obviously when you use that parameter, things are delicate and you have to watch if you have any names repeated / ambiguous / etc.

**Reply by Dreamsorcerer** [reply]
That certainly sounds right, but I still can't seem to get it to work:
```
cols = (sa.case((RhymeUS.id != None, getattr(RhymeUS, c.name)), else_=getattr(Rhyme, c.name)).label(c.name) for c in Rhyme.__table__.columns)
sub_query = sa.select(*cols).outerjoin(RhymeUS).select_from(Rhyme).subquery()
alias = aliased(Rhyme, alias=sub_query, adapt_on_names=True)
where = sa.and_(sub_query.c.ph1 == ph1, sub_query.c.syl == 1, sub_query.c.aux_syl == 0)
stmt = sa.select(alias).from_statement(sa.select(sub_query).where(where))
```
On the last line I've tried Rhyme/alias/sub_query in the first select(), and alias/sub_query in the second select(), and some other things. It still always returns the values from the rhymes table, not rhymes_us (though the filtering is working on the rhymes_us table).

Is there something I'm doing obviously wrong?

**Reply by zzzeek** [reply]
> Is there something I'm doing obviously wrong?

i can't tell with a code fragment I'd need something I can run so I can see what names are rendered and how they are matching, then if that's not working I'd have to dig into aliased() to see if there are some assumptions that I'm not recalling

is there a reason the working text() version isn't sufficient, if you just use the [.columns()](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.TextClause.columns) method you can link ORM columns positionally as in [this example](https://docs.sqlalchemy.org/en/20/orm/queryguide/select.html#getting-orm-results-from-textual-statements)

**Reply by Dreamsorcerer** [reply]
Could do, it'd just be nicer/safer to do it without raw SQL.

> I'd need something I can run

I'm not sure what's going on now. I created a minimal reproducer, but it works fine. The code looks identical and produces the same SQL...

**Reply by zzzeek** [reply]
OK maybe caching issue ?    try sending https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine.params.query_cache_size = 0 , make sure that's not the problem

**Reply by zzzeek** [reply]
note the caching is involved all throughout how rows are matched to the ORM objects and such , it's not just the SQL string

**Reply by Dreamsorcerer** [reply]
Yes, just figured it out, seems to be something to do with object caching in the session.

If I add a `r = await sess.get(Rhyme, 1)` before the main query, then the issue appears. Not sure if that counts as a bug in sqlalchemy, but certainly surprising.

Here's a full reproducer:
```py
import asyncio
import sqlalchemy as sa
from sqlalchemy import MetaData
from sqlalchemy.ext.asyncio import create_async_engine, async_sessionmaker
from sqlalchemy.orm import DeclarativeBase, Mapped, aliased, mapped_column, relationship

class Base(DeclarativeBase):
    pass

@sa.orm.declarative_mixin
class _Rhyme:
    ph1: Mapped[int]

class Rhyme(Base, _Rhyme):
    __tablename__ = "rhymes"

    id: Mapped[int] = mapped_column(primary_key=True)

    rhyme_us: Mapped["RhymeUS"] = relationship("RhymeUS", cascade="save-update, merge, delete, delete-orphan")

class RhymeUS(Base, _Rhyme):
    __tablename__ = "rhymes_us"

    id: Mapped[int] = mapped_column(sa.ForeignKey(Rhyme.id, ondelete="CASCADE"), primary_key=True)

async def main():
    engine = create_async_engine("sqlite+aiosqlite:///:memory:")
    db = async_sessionmaker(engine, expire_on_commit=False)

    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)
    async with db.begin() as sess:
        r = Rhyme(ph1=15)
        sess.add(r)
    async with db.begin() as sess:
        sess.add(RhymeUS(id=r.id, ph1=10))

    async with db() as sess:
        r = await sess.get(Rhyme, 1)  # <--- This breaks the result of the next query

        cols = (sa.case((RhymeUS.id != None, getattr(RhymeUS, c.name)), else_=getattr(Rhyme, c.name)).label(c.name) for c in Rhyme.__table__.columns)
        sub_query = sa.select(*cols).outerjoin(RhymeUS).select_from(Rhyme).subquery()
        alias = aliased(Rhyme, alias=sub_query, adapt_on_names=True)
        stmt = sa.select(alias).from_statement(sa.select(sub_query).where(sub_query.c.ph1 == 10))
        results = await sess.scalars(stmt)
        rhymes = results.all()

    assert rhymes[0].ph1 == 10

asyncio.run(main())
```

**Reply by zzzeek** [reply]
if your second query includes a row for Rhyme with a primary key of 1 and the other values are not the same as what they are in the primary rhyme table, that's expected.   Use [populate_existing](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#orm-queryguide-populate-existing) anytime an ORM query expect to re-interpret attributes of an ORM object for the same primary key identity.
